outline.md

```diff
@@ -16,30 +61,127 @@
+  * `activate`: perform [*environment modifications*](#environment-modifications) for the [concretized](#concretization) version of all abstract specs provided.
+    * This corresponds to [*global activation*][^global-activation] specifically.
+    * This abstract action covers the act of *creating a non-global filesystem view* as well.
+    * The input here may come as a *module*.
+* Spack also has many other commands which affect the output of these commands, but WOLOG they can be ignored.
+
+#### Spec syntax
+* Syntax for constraints on Spec DAGs on command line
+* Query specs, specs as "matchers"
 
-| type  | availability                                       | components                           | propagates | 
+<!-- **A spec is an index into an "environment"** -->
+
+#### Build Phases
+* Spack defines several *build phases* which may or may not execute for some [package](#package), depending on the user-selected spack command.
+  * *See [supported spack workflows](https://spack.readthedocs.io/en/latest/workflows.html).*
+
+| type  | availability                                       | components                           | propagates |
 | ---   | ---                                                | ---                                  | ---        |
 | build | build time, not runtime                            | Unknown                              | no         |
 | link  | build time, also implies run and test time for now | Unknown, probably includes a library | yes        |
 | run   | all of build, link run and test                    | Unknown                              | yes        |
 | test  | test build and run time, maybe normal build        | Unknown                              | no         |
 
-### Things not accounted for:
+* Each *build phase* acts upon a [unified environment](#unified-environment), and consists of:
+  1. applying the appropriate [*environment modifications*](#environment-modifications).
+  1. arbitrary work that modifies the *filesystem*.
 
-* PATHs required by a given phase
-* Components used, probably by phase:
-  * commands? might be covered by paths
-  * interface: headers/modules/definitions that are needed at build but not runtime
-  * libraries: possibly also static vs dynamic
-* propagation: in A->B->C does A depend on C in a given phase
-* Does it have ABI impact?
+##### Build Plan
+* A *build plan* is an ordered list of *build phases*.
+
+### Environments
+*From [Environments](https://spack.readthedocs.io/en/latest/environments.html):*
+> An environment is used to group together a set of specs for the purpose of building, rebuilding and deploying in a coherent fashion.
+
+* A *spack environment*
+* What are they?
+  * How/Can these be serialized?
+  * How do these differ from (spack) modules?
+* Shell session?
+* Shared env vars with common parent process?
+
+## Package Developer Concepts
+
+### Package
```

Comment:
This section is good and probably needs to be moved up front -- we need to define packages, the package graph, and differences between that and the graph of libraries *within* packages.

Reply:
That's exactly why I chose to develop it by user, then package developer, then maintainer. One alternative is to do this the complete opposite direction by describing the spack implementation, then the package interface, then the user interface, but that doesn't give us the opportunity to demonstrate that our model is fully general and not something arbitrary.

Reply:
If moving up front is due to ordering by importance instead of by concept development, then I think that would make sense, as there's enough back edges in this tree now that it would be fine. I also think that pulling out the actual workflows that a user vs a package developer would expect and making it part of the initial Use Cases section would make sense.